services/remotecluster/send.go

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package remotecluster
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"github.com/mattermost/mattermost-server/v5/mlog"
+	"github.com/mattermost/mattermost-server/v5/model"
+)
+
+type SendResultFunc func(msg model.RemoteClusterMsg, remote *model.RemoteCluster, resp []byte, err error)
+
+type sendTask struct {
+	msg model.RemoteClusterMsg
+	f   SendResultFunc
+}
+
+// SendOutgoingMsg sends a message to all remote clusters interested in the message's topic.
+//
+// `ctx` determines behaviour when the outbound queue is full. A timeout or deadline context will return a
+// BufferFullError if the message cannot be enqueued before the timeout. A background context will block indefinitely.
+//
+// An optional callback can be provided that receives the success or fail result of sending to each remote cluster.
+// Success or fail is regarding message delivery only.  If a callback is provided it should return quickly.
+func (rcs *Service) SendOutgoingMsg(ctx context.Context, msg model.RemoteClusterMsg, f SendResultFunc) error {
+	task := sendTask{
+		msg: msg,
+		f:   f,
+	}
+
+	if ctx == nil {
+		ctx = context.Background()
+	}
+
+	select {
+	case rcs.send <- task:
+		return nil
+	case <-ctx.Done():
+		return NewBufferFullError(cap(rcs.send))
+	}
```

Comment:
A context could also be canceled in which case I am not sure returning this error would be very clear.
Maybe we could wrap `ctx.Err()` to be more informative on the cause of failure.

Reply:
Fixed via https://github.com/mattermost/mattermost-server/pull/16432/commits/c785b4b692a65411487822c2ce7c40f614bdd121